### PR TITLE
Core/data communicator serial defaults

### LINKED ISSUE #3646: What is a reasonable serial default for DataCommunicator functions?
The question at hand is what to do with the two-buffer wrappers for MPI calls. Given the debate around this point in #3340 and in the technical committee, I wanted to open a separate issue on this specific topic. As currently implemented in #3340, functions such as
https://github.com/KratosMultiphysics/Kratos/blob/7878279d474f6f9288b098bf0c977988c7782a7b/kratos/includes/data_communicator.h#L307-L311
Are simply empty in the serial version, so the output buffer is left untouched. This is in contrast to the return buffer version
https://github.com/KratosMultiphysics/Kratos/blob/7878279d474f6f9288b098bf0c977988c7782a7b/kratos/includes/data_communicator.h#L284-L287
where something is always returned.

I am not too happy with the current design, mainly because it is inconsistent (functions that do one thing in serial and a different one in parallel). I see different options going forward, but I'd like to hear the opinion of the eventual users of this class and the @KratosMultiphysics/design-committee in particular:
- Leave things as they are now. The behavior is inconsistent, but documented by tests, and my recommendation would be to only use the two-buffer functions within `if (rComm.IsDistributied())` checks.
- Return some reasonable value (for example, for reduction-based methods, the local value) as output. This introduces a subtler problem: for some methods no reasonable return value exists (send-receive if not sending to rank 0, for example). In such cases, it might make sense to return an empty buffer (the return buffer version does so), but this goes against the rule never resizing the buffer, which the mpi version follows strictly.
- Just remove the two-buffer methods (but they are in general more efficient if the user knows what they are doing).
- Return a reasonable value if possible, throw an error otherwise. In practice, this means that functions that don't have an obvious return value in serial can only be used within `if (rComm.IsDistributied())` checks.

What do you think? FYI @KratosMultiphysics/design-committee @RiccardoRossi @pooyan-dadvand @roigcarlo @adityaghantasala.

### PR DESCRIPTION
**Pull request by jcotela**
Improving the default behavior of the DataCommunicator in non-MPI runs to be consistent with the MPI implementation. The idea is that DataCommunicator methods should behave as the MPI equivalents if those were run with a single process. This allows MPI-parallel algorithms that "reduce" nicely to the single-process case to work independently of the specifics of the run and the type of DataCommunicator (serial/distributed).
In short, everything should work nice as long as the single-process case never explicitly requires out-of-process data. If out-of-process data is explicitly requested, the method will throw an error.

Tests have been updated accordingly.

This closes #3646

### CODE REVIEW

**Comment by philbucher** [comment]
from my side this is fine, it is what was previously discussed

I will approve in some days if none else does.

Thanks @jcotela

**Review by msandre — APPROVED** [review]
Since there have been no further comments, I am approving. Thx :+1:

**Comment by jcotela** [comment]
Thanks!
